core/src/main/java/org/elasticsearch/discovery/single/SingleNodeDiscovery.java

```diff
@@ -0,0 +1,142 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A discovery implementation where the only member of the cluster is the local node.
+ */
+public class SingleNodeDiscovery extends AbstractLifecycleComponent implements Discovery {
+
+    private final ClusterService clusterService;
+    private final DiscoverySettings discoverySettings;
+
+    public SingleNodeDiscovery(final Settings settings, final ClusterService clusterService) {
+        super(Objects.requireNonNull(settings));
+        this.clusterService = Objects.requireNonNull(clusterService);
+        final ClusterSettings clusterSettings =
+                Objects.requireNonNull(clusterService.getClusterSettings());
+        this.discoverySettings = new DiscoverySettings(settings, clusterSettings);
+    }
+
+    @Override
+    public DiscoveryNode localNode() {
+        return clusterService.localNode();
+    }
+
+    @Override
+    public String nodeDescription() {
+        return clusterService.getClusterName().value() + "/" + clusterService.localNode().getId();
+    }
+
+    @Override
+    public void setAllocationService(final AllocationService allocationService) {
+
+    }
+
+    @Override
+    public void publish(final ClusterChangedEvent event, final AckListener listener) {
+
+    }
+
+    @Override
+    public DiscoveryStats stats() {
+        return null;
+    }
+
+    @Override
+    public DiscoverySettings getDiscoverySettings() {
+        return discoverySettings;
+    }
+
+    @Override
+    public void startInitialJoin() {
+        final ClusterStateTaskExecutor<DiscoveryNode> executor =
+                new ClusterStateTaskExecutor<DiscoveryNode>() {
+
+                    @Override
+                    public ClusterTasksResult<DiscoveryNode> execute(
+                            final ClusterState current,
+                            final List<DiscoveryNode> tasks) throws Exception {
+                        assert tasks.size() == 1;
+                        final DiscoveryNodes.Builder nodes =
+                                DiscoveryNodes.builder(current.nodes());
+                        // always set the local node as master, there will not be other nodes
+                        nodes.masterNodeId(localNode().getId());
+                        final ClusterState next =
+                                ClusterState.builder(current).nodes(nodes).build();
+                        final ClusterTasksResult.Builder<DiscoveryNode> result =
+                                ClusterTasksResult.builder();
+                        return result.successes(tasks).build(next);
+                    }
+
+                    @Override
+                    public boolean runOnlyOnMaster() {
+                        return false;
+                    }
+
+                };
+        final ClusterStateTaskConfig config = ClusterStateTaskConfig.build(Priority.URGENT);
+        clusterService.submitStateUpdateTasks(
```

Comment:
No, it wasn't my fingers. I see why I did this. A `ClusterStateUpdateTask` can not override `runOnlyOnMaster` which we need to set to `false` here. 😄